### ISSUE TITLE
Feature Request: X11 systray support

### ISSUE THREAD

**Issue by plomari** [dunst-project/dunst#1036]
Dunst could support the classic X11 systray (https://specifications.freedesktop.org/systemtray-spec/systemtray-spec-0.3.html), and show visual feedback of the current status in it. The most important status would be whether messages have been "missed". This status would get cleared by clicking on the systray icon, or whether messages were explicitly closed.

**Comment by fwsmit** [comment]
Hi plomari,
I support the idea of a systray application. I think, however, that it would be best suited for a separate application. Dunst already supports most of the neccesary functionality via dbus. Take a look at the man page for dunstctl for a summary of most of the functions supported via dbus. The main things that would need to be implemented is a way to signal which notifications have been missed and to clear this flag.

By doing it in a separate application you could support any number of notification daemons without much effort.

**Comment by plomari** [comment]
It likely requires some hairy low level X11 code and more, so I thought it'd be ideal for the dunst project to implement, somehow.

**Comment by fwsmit** [comment]
It's the same amount of X11 code when it's a separate project. There are probably libraries that make it easier, but again, I'll leave it to some other project to pick up.

**Comment by plomari** [comment]
Also, what doesn't make this easier, is that dbus is very hard to use.

**Comment by fwsmit** [comment]
No it's not so hard. Otherwise you could use dunstctl as an intermediate. I'm still not convinced this should be part of dunst

**Comment by plomari** [comment]
Another aspect that makes an external implementation quite inconvenient is that it needs to poll for new notifications.

**Comment by fwsmit** [comment]
> Another aspect that makes an external implementation quite inconvenient is that it needs to poll for new notifications.

No, you can subscribe to dbus, which gives you a ping when a new notification arrives.

**Comment by plomari** [comment]
But then I have to use the incredibly complicated dbus AP (instead of dunstctl), and then I won't need dunst either.

**Comment by fwsmit** [comment]
Are you planning on implementing this yourself?

**Comment by fwsmit** [comment]
> Implement what exactly?

A X11 systray utility for dunst.

> Currently I'm pondering multiple possibilities. The easiest would be whipping up something with Python and GTK, and making the window icon a systray icon with `alltray` or similar.

Alltray is unmaintained, so I wouldn't recommend that.

After considering it a bit more, I think systray support for dunst could be interesting. I wouldn't rule it out completely. But it shouldn't add big dependencies and should also work on wayland. But still, it's probably easier to implement as a separate application. I would be happy to add any neccesary dbus interfaces (really, dbus isn't that hard, take a look at dunstctl).

> But this has questionable long-term stability and reduced interaction through systray.

What do you mean? What would not be stable?

**Comment by plomari** [comment]
> Alltray is unmaintained, so I wouldn't recommend that.

It'll work forever. But I use IceWM, which doesn't even need alltray.

> After considering it a bit more, I think systray support for dunst could be interesting. I wouldn't rule it out completely.

Nice.

> But it shouldn't add big dependencies and should also work on wayland.

X11 systray is just a tiny low level protocol on top of X11 mechanisms, and it doesn't even need any libraries. I don't know if Wayland, I expect every DE has its own custom protocol.

When I first found dunst, it actually used raw X11 and the dependency-free raw libdbus libraries as only dependencies. Now it depends on a number of libraries, even parts of Gnome (gobject, pango), which makes it fairly big. I don't think systray support would make it significantly heavier.

> But still, it's probably easier to implement as a separate application.

There are some good reasons you brought up why it should be a separate application. But I disagree that it'd be easier.

> (really, dbus isn't that hard, take a look at dunstctl).

As someone who messed with dbus both on the lowest level (wire protocol) and the highest level (automagic Python dbus bindings), I can't agree at all. Your average web API is a marvel of simplicity compared to anything related to dbus.

**Comment by plomari** [comment]
> No, you can subscribe to dbus, which gives you a ping when a new notification arrives.

In addition to being excruciatingly complex, a worse problem is that apparently only 1 process can listen to 'org.freedesktop.Notifications', so only 1 process can receive notifications, so this can't be implemented outside of dunst.

I really wonder why the "Linux desktop" is still using such a horrible technology (extremely complex and inflexible) as dbus. It makes web technology look simple and elegant.

**Comment by fwsmit** [comment]
> In addition to being excruciatingly complex, a worse problem is that apparently only 1 process can listen to 'org.freedesktop.Notifications'

I don't see why not. I tried running `dbus-monitor path=/org/freedesktop/Notifications` in two different terminals and that works just fine. Anyways, I'm gonna close this issue, since it doesn't seem like a good idea to implement inside of dunst.
